### DISCUSSION TITLE
PRI file format docs?

### DISCUSSION THREAD

**Discussion by mikehearn** [microsoft/WindowsAppSDK#2828] [Q&A] [answered]
Hi,

@axelandrejs Are there any docs anywhere about the format of PRI files? I can see that there's extensive code in this repository for loading from them, but I couldn't find any code comments or files where the format itself is documented. And it seems that the code here is for _reading_ PRI files but not _making_ them - is that understanding correct?

My goal is to generate a PRI file that enables the use of unplated icon assets, as Windows doesn't seem to use plated icons anymore but will still do plating by default if you don't have a PRI file. And I'd like to do it without using any Win32 or WinRT APIs.

**Comment by axelandrejs** [comment, marked as answer]
Hey!

The format is not formally documented. The code to write the format is actually mostly present in MRT core - makepri.exe mostly just adds code to read the source data.

That said, you don't need your own file writer to achieve what you want. How the various qualifiers, including plating are specified is described here: https://docs.microsoft.com/en-us/windows/uwp/app-resources/tailor-resources-lang-scale-contrast. 

To achieve what you want, the best way is probably to have an asset with altform-unplated as well as one with no altform at all in your project. Makepri.exe is standalone so you can run it outside of VS or any other build system.

**Reply by mikehearn** [reply]
Thanks for the response! The reason I need to write the format myself is that I run a company that makes a cross-platform packaging tool. Check it out, it's called [Hydraulic Conveyor](https://www.hydraulic.software). The tool is intended to make distributing cross-platform desktop apps as easy as web apps, and part of what makes it so easy is that it can build and sign every package type on every OS i.e. MSIX packages can be built on macOS and Linux, because it uses its own code for everything.

And it works! Except, the icons end up "plated" which looks very odd. Our investigation concluded that the Visual Studio toolchain actually overrides the plating system and disables it by supplingy unplated assets _by default_, so plated icons aren't really used for any MS first party apps anymore. Maybe some legacy of the Windows Phone days. That's why our packages look a bit odd. But to supply unplated assets you can't just stick them in the MSIX package at special paths. You pick whatever paths you want, and then those paths are embedded into a PRI file, which is at a special path. Hence, we think we have to add support for generating PRI files.

Now, the question is what's the best way to do that _without_ using Win32 or makepri? It seems that because the PRI file really just contains file pointers it may be possible to just create a sort of 'template' PRI that points at fixed/hard coded image paths but this doesn't work because the PRI contains the name of the package inside it for some reason. So one-PRI-for-all doesn't quite work. _But_ from binary diffing, it seems like it should be possible to just edit the name and re-adjust the rest of the file to match. Then the PRI should be enough to disable plating.

Of course that's very hacky and not easy to understand. That's why I was wondering if there's any format docs. Seems the best bet to understand what we have to do is read the MrtCore code? Can you suggest where to start, as it's a fairly big codebase? Alternatively, can you think of a better way we can make our packages use unplated assets?

I've also looked a bit at compiling MrtCore for macOS and Linux, then we could ship it with the app and use it to make PRI files cross platform. But it looks like it's a Win32 codebase and really does depend on the OS in a bunch of ways for COM and the like, so that's probably infeasible outside of a really heavy approach like Wine.

**Comment by axelandrejs** [comment]
It should actually be doable to compile MRT core for other platforms. MRT core consists of three layers.

The core runtime
The flat-C API over that core runtime
The WinRT API over the flat-C API

You can find the core runtime here: https://github.com/microsoft/WindowsAppSDK/tree/main/dev/MRTCore/mrt/mrm. The part related to building a PRI file is here: https://github.com/microsoft/WindowsAppSDK/tree/main/dev/MRTCore/mrt/mrm/include/mrm/build

This code abstracts away platform calls to make the code portable. See https://github.com/microsoft/WindowsAppSDK/blob/main/dev/MRTCore/mrt/mrm/include/mrm/common/Platform.h. EG, we use _defcreatefile as an abstraction over CreateFile. If you implement this abstraction in your other target OS it should mostly work. I say mostly because I think in a few places direct calls might have snuck in. So there might be a bit of work. But principally, MRT core is meant to be portable at the foundational layer, which is all you need.

**Reply by mikehearn** [reply]
Ah, that's great news! I also see there's some info on the basic structures in `MrmFile.h`

That should be enough to get us started. Can I just double check quickly that our understanding is correct, that the only way to make the icon border plating go away is with a PRI file? There's no other way to supply these image in an MSIX, other than going through MRM?

**Comment by axelandrejs** [comment]
I don't actually know the details about the shell contract. @ptorr-msft do you know?

**Reply by ptorr-msft** [reply]
No; I can try and find out internally though. @mikehearn how are you dealing with localization etc. if you're not using a PRI file? I don't know how you'd get the Shell to have localized app names etc. without a PRI file?

**Reply by mikehearn** [reply]
The apps are cross platform so don't really use MRM themselves - the PRI is only for package metadata. Indeed, without a PRI we (probably?) can't localize the package display name, but that's a lower priority compared to the icons issue which is immediately visible to everyone. And most package/app names aren't localized anyway (brand name+product name), especially for internal or specialist apps.

For now we're looking for something that can be done quickly to remove the plating. Using more of the MSIX / PRI features can come later.
